Add grid mask for regional domains to speed up nearest n. interpolation

DusanJovic-NOAA opened the pull request:
## Description

This PR adds grid mask to regional output domains, with values indicating whether the output grid point is outside the forecast computational domain. When this mask is present in the esmf fields, the computation of nearest neighbor route handles is much faster, especially if the region of the output grid that is outside the computational domain is large.


### Issue(s) addressed

Link the issues to be closed with this PR, whether in this repository, or in another repository.
(Remember, issues should always be created before starting work on a PR branch!)
- fixes #987

## Testing

How were these changes tested?  UFS regression test on Ursa
What compilers / HPCs was it tested with?  Intel and GNU on Ursa.
Are the changes covered by regression tests? Yes.
Have the ufs-weather-model regression test been run? Yes.
On what platform?  Ursa
- Will the code updates change regression test baseline? No.
- Please commit the regression test log files in your ufs-weather-model branch


## Dependencies

If testing this branch requires non-default branches in other repositories, list them.
Those branches should have matching names (ideally)

Do PRs in upstream repositories need to be merged first?
If so add the "waiting for other repos" label and list the upstream PRs
- waiting on noaa-emc/nems/pull/<pr_number>
- waiting on noaa-emc/fv3atm/pull/<pr_number>

DusanJovic-NOAA committed:
Add destination mask for regional history output grids

DusanJovic-NOAA committed:
Merge remote-tracking branch 'origin/develop' into reg_grid_mask

DusanJovic-NOAA committed:
Merge remote-tracking branch 'origin/develop' into reg_grid_mask

DusanJovic-NOAA committed:
Code cleanup

DusanJovic-NOAA committed:
Merge remote-tracking branch 'origin/develop' into reg_grid_mask

DusanJovic-NOAA committed:
Code cleanup

DusanJovic-NOAA committed:
Code cleanup

DusanJovic-NOAA committed:
Merge remote-tracking branch 'origin/develop' into reg_grid_mask

BinLiu-NOAA reviewed (approved):
Thanks for enabling this capability, which can substantially speed up the write grid component initialization phase for high-resolution regional configurations!

junwang-noaa commented on fv3/io/module_wrt_grid_comp.F90:
```diff
@@ -4761,6 +4510,244 @@ subroutine compute_fields_checksum(bundle, rc)
+          do jg=jstart, jend
+            do ig=istart, iend
+              if (ptr(ig,jg) == 0.0 .and. ptr_dst_status(ig,jg) /= ESMF_REGRIDSTATUS_OUTSIDE) then
+                write(0,*)'remap warning: ', ig,jg,ptr(ig,jg), ptr_dst_status(ig,jg)
```
I am not sure if we need to keep this section, if this happens, then something is wrong with ESMF.
Maybe more details: "Remap warning - not mapped but not outside source grid: "
line 4698: "Remap warning - mapped but ESMF status does not show mapped: "

DusanJovic-NOAA replied:
Yes, this section is not needed as long as everything works as we expect, I just used it to double-check that all points outside and only points outside computational domain are zero, and that all others are non-zero. I can either remove this entire section or update the warning prints, what do you prefer?

junwang-noaa replied:
I prefer removing it.

DusanJovic-NOAA replied:
Removed in 648b545

junwang-noaa commented on fv3/fv3_cap.F90:
```diff
@@ -1266,6 +1356,7 @@ subroutine ModelAdvance_phase2(gcomp, rc)
           if (fieldCount > 0) then
             call ESMF_FieldBundleSMM(fcstFB(j), wrtFB(j,n_group),         &
                                      routehandle=routehandle(j, n_group), &
+                                     zeroregion=ESMF_REGION_SELECT,       &
```
Is this required if the ESMF_FieldBundleRegridStore has a dstMaskValues mask? I saw in ESMF document (https://earthsystemmodeling.org/docs/nightly/v8.9.0b11/ESMF_refdoc/node5.html#SECTION050253300000000000000):

type(ESMF_Region_Flag), intent(in), optional :: zeroregion ! DEPRECATED ARGUMENT

DusanJovic-NOAA replied:
Yes, this flag is needed for moving nest. I replaced it with `zeroregionflag` in 4e82ca2

DusanJovic-NOAA committed:
Remove 'dstStatusField'

Used to double-check that the destination status and remapped values are consistent

DusanJovic-NOAA committed:
Replace 'zeroregion' with 'zeroregionflag' in ESMF_FieldBundleSMM call

This argument ('zeroregion') has been deprecated since ESMF v8.1.0. Use
'zeroregionflag' instead.

DusanJovic-NOAA committed:
Merge remote-tracking branch 'origin/develop' into reg_grid_mask

DusanJovic-NOAA committed:
Merge remote-tracking branch 'origin/develop' into reg_grid_mask
